crates/shared/src/sources/uniswap_v3/event_fetching.rs

```diff
@@ -0,0 +1,159 @@
+        let standard_event: Option<Result<UniswapV3Event, ExecutionError>> =
+            log.topics.get(0).copied().map(|topic| match topic {
+                H256(hex!("0c396cd989a39f4459b5fa1aed6a9a8dcdbc45908acfd67e028cd568da98982c")) => {
+                    Ok(UniswapV3Event::Burn(
+                        log.clone().decode(
+                            UniswapV3Pool::raw_contract()
+                                .abi
+                                .event("Burn")
+                                .expect("generated event decode"),
+                        )?,
+                    ))
+                }
+                H256(hex!("7a53080ba414158be7ec69b987b5fb7d07dee101fe85488f0853ae16239d0bde")) => {
+                    Ok(UniswapV3Event::Mint(
+                        log.clone().decode(
+                            UniswapV3Pool::raw_contract()
+                                .abi
+                                .event("Mint")
+                                .expect("generated event decode"),
+                        )?,
+                    ))
+                }
+                H256(hex!("c42079f94a6350d7e6235f29174924f928cc2ac818eb64fed8004e115fbcca67")) => {
+                    Ok(UniswapV3Event::Swap(
+                        log.clone().decode(
+                            UniswapV3Pool::raw_contract()
+                                .abi
+                                .event("Swap")
+                                .expect("generated event decode"),
+                        )?,
+                    ))
+                }
+                _ => Err(ExecutionError::from(Error::InvalidData)),
+            });
+        if let Some(Ok(data)) = standard_event {
+            return Ok(data);
+        }
+        Err(ExecutionError::from(Error::InvalidData))
+    }
+}
+
+pub struct UniswapV3PoolEventFetcher(pub Web3);
+
+impl EventRetrieving for UniswapV3PoolEventFetcher {
+    type Event = UniswapV3Event;
+    fn get_events(&self) -> DynAllEventsBuilder<Self::Event> {
+        let mut events = DynAllEventsBuilder::new(self.0.clone(), H160::default(), None);
+        let events_signatures = vec![
+            H256::from_str(
+                "0xc42079f94a6350d7e6235f29174924f928cc2ac818eb64fed8004e115fbcca67", //swap
+            )
+            .unwrap(),
+            H256::from_str(
+                "0x0c396cd989a39f4459b5fa1aed6a9a8dcdbc45908acfd67e028cd568da98982c", //burn
+            )
+            .unwrap(),
+            H256::from_str(
+                "0x7a53080ba414158be7ec69b987b5fb7d07dee101fe85488f0853ae16239d0bde", //mint
+            )
+            .unwrap(),
+        ];
+        events.filter = events
+            .filter
+            .address(vec![])
+            .topic0(events_signatures.into());
+        events
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct RecentEventsCache {
+    /// Events are ordered by block number
+    events: Vec<Event<UniswapV3Event>>,
```

Comment:
Wondering if I should use `BTreeMap` with block numbers as keys, instead of `Vec`.
Theoretically, `BTreeMap` should probably be faster for retrieving range of events when `events` is large.

Reply:
For UniswapV3, I've noticed around 5 events per block making it ~35k events daily if I am not wrong... Maybe it would be easier to add the option to cap the max number of blocks in event storage. For liquidity fetchers I don't see a point of saving the full history of events from app startup. If we would ever want to enable history liquidity fetching (for some analysis for example) then we would have to move the storage to db anyway

Reply:
BTreeMap fits the semantics of this struct nicely where you delete and retrieve events in ranges. I'd use it.

Reply:
Ended up using `BTreeMap`. Appending is probably a bit slower but fetching should be much faster.